File: app/src/main/java/cn/blinkdagger/androidLab/widget/CustomPopupWindow.java
package cn.blinkdagger.androidLab.widget;

import android.app.Activity;
import android.graphics.drawable.ColorDrawable;
import androidx.annotation.AnimatorRes;
import androidx.annotation.FloatRange;
import androidx.annotation.LayoutRes;
import androidx.annotation.StyleRes;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.PopupWindow;

/**
 * 类描述：通用PopupWindow
 * 创建人：ls
 * 创建时间：2017/12/7
 * 修改人：
 * 修改时间：
 * 修改备注：
 */

public class CustomPopupWindow extends PopupWindow {

    private Activity activity;
    private float onShowBgAlpha;//显示的时候背景的透明度
    private boolean onShowChangeBg;//显示的时候背景透明度是否变化
    private boolean onDissmisChangeBg;//消失的时候背景透明度是否变化


    private CustomPopupWindow(View contentView, int width, int height) {
        super(contentView, width, height);
    }

    public void setActivity(Activity activity) {
        this.activity = activity;
    }

    public void setOnShowBgAlpha(@FloatRange(from = 0f, to = 1.0f) float onShowBgAlpha) {
        this.onShowBgAlpha = onShowBgAlpha;
    }

    public void setOnShowChangeBg(boolean onShowChangeBg) {
        this.onShowChangeBg = onShowChangeBg;
    }

    public void setOnDissmisChangeBg(boolean onDissmisChangeBg) {
        this.onDissmisChangeBg = onDissmisChangeBg;
    }

    public void showChangeBg() {
        if (onShowChangeBg) {
            if (activity != null) {
                setBackgroundAlpha(activity, onShowBgAlpha);
            }
        }
    }

    @Override
    public void showAtLocation(View parent, int gravity, int x, int y) {
        showChangeBg();
        super.showAtLocation(parent, gravity, x, y);
    }

    @Override
    public void showAsDropDown(View anchor, int xoff, int yoff, int gravity) {
        super.showAsDropDown(anchor, xoff, yoff, gravity);
        showChangeBg();
    }

    @Override
    public void dismiss() {
        if (onDissmisChangeBg) {
            if (activity != null) {
                setBackgroundAlpha(activity, 1.0f);
            }
        }
        super.dismiss();
    }

    public static class Builder {

        private Activity activity;
        private View contentView;// 内容视图
        private int width = ViewGroup.LayoutParams.WRAP_CONTENT;//宽度
        private int height = ViewGroup.LayoutParams.WRAP_CONTENT;//高度
        private @StyleRes int anmationStyle;// 动画
        private boolean focusable;//是否可以获取焦点
        private boolean outsideTouchable;//外部是否可以点击
        private boolean onShowChangeBg = true; //显示的时候背景透明度是否变透明【默认变化】
        private boolean onDissmisChangeBg = true;//消失的时候背景透明度是否还原【默认还原】
        private float onShowBgAlpha = 0.5f;//显示的时候背景的透明度
        private boolean onKeyBackDissmiss = false;//是否拦截返回键消失【默认不拦截】
        private OnDismissListener onDismissListener;// 消失监听事件

        public Builder(Activity activity) {
            this.activity = activity;
        }

        public Builder setContentView(View contentView) {
            this.contentView = contentView;
            return this;
        }

        public Builder setContentView(@LayoutRes int contentViewId) {
            this.contentView = activity.getLayoutInflater().inflate(contentViewId, null);
            return this;
        }

        public Builder setContentView(View contentView, int width, int height) {
            this.width = width;
            this.height = height;
            return this;
        }

        public Builder setContentView(@LayoutRes int contentViewId, int width, int height) {
            this.contentView = activity.getLayoutInflater().inflate(contentViewId, null);
            this.width = width;
            this.height = height;
            return this;
        }

        public Builder setAnmationStyle(@AnimatorRes int anmationStyle) {
            this.anmationStyle = anmationStyle;
            return this;
        }

        public Builder setFocusable(boolean focusable) {
            this.focusable = focusable;
            return this;
        }

        public Builder setOutsideTouchable(boolean outsideTouchable) {
            this.outsideTouchable = outsideTouchable;
            return this;
        }

        public Builder setOnShowChangeBg(boolean onShowChangeBg) {
            this.onShowChangeBg = onShowChangeBg;
            this.focusable = true;
            return this;
        }

        public Builder setOnDissmisChangeBg(boolean onDissmisChangeBg) {
            this.onDissmisChangeBg = onDissmisChangeBg;
            return this;
        }

        public Builder setOnShowBgAlpha(@FloatRange(from = 0f, to = 255) float onShowBgAlpha) {
            this.onShowBgAlpha = onShowBgAlpha;
            return this;
        }

        public void setOnDismissListener(OnDismissListener onDismissListener) {
            this.onDismissListener = onDismissListener;
        }

        /**
         * 设置点击外部本身是否消失
         *
         * @param onOutsideTouchDissmiss
         */
        public Builder setOnOutsideTouchDissmiss(boolean onOutsideTouchDissmiss) {
            if (onOutsideTouchDissmiss) {
                this.outsideTouchable = true;
                this.focusable = true;
            } else {
                this.outsideTouchable = false;
                this.focusable = false;
            }
            return this;
        }

        /**
         * 设置点击返回键本身是否消失
         *
         * @return
         */
        public Builder setOnKeyBackDissmiss(boolean onKeyBackDissmiss) {
            this.onKeyBackDissmiss = onKeyBackDissmiss;
            return this;
        }

        public CustomPopupWindow build() {
            if (contentView == null) {
                throw new NullPointerException("contentView con't be null ");
            }
            final CustomPopupWindow popupWindow = new CustomPopupWindow(contentView, width, height);
            popupWindow.setActivity(this.activity);
            popupWindow.setOutsideTouchable(this.outsideTouchable);
            popupWindow.setFocusable(this.focusable);
            popupWindow.setOnDissmisChangeBg(this.onDissmisChangeBg);
            popupWindow.setOnShowChangeBg(this.onShowChangeBg);
            popupWindow.setOnShowBgAlpha(this.onShowBgAlpha);
            popupWindow.setOnDismissListener(this.onDismissListener);
            popupWindow.setBackgroundDrawable(new ColorDrawable());
            if (onKeyBackDissmiss) {
                this.contentView.setFocusable(true); // 这个很重要
                this.contentView.setFocusableInTouchMode(true);
                this.contentView.setOnKeyListener(new View.OnKeyListener() {
                    @Override
                    public boolean onKey(View v, int keyCode, KeyEvent event) {
                        if (keyCode == KeyEvent.KEYCODE_BACK) {
                            popupWindow.dismiss();
                            return true;
                        }
                        return false;
                    }
                });
            }

            return popupWindow;
        }
    }

    /**
     * 设置页面的透明度
     *
     * @param bgAlpha 1表示不透明
     */
    private void setBackgroundAlpha(Activity activity, float bgAlpha) {
        WindowManager.LayoutParams lp = activity.getWindow().getAttributes();
        lp.alpha = bgAlpha;
        if (bgAlpha == 1) {
            activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND);//不移除该Flag的话,在有视频的页面上的视频会出现黑屏的bug
        } else {
            activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND);//此行代码主要是解决在华为手机上半透明效果无效的bug
        }
        activity.getWindow().setAttributes(lp);

    }
}

File: app/src/main/java/cn/blinkdagger/androidLab/widget/CustomPopupWindow.kt
package cn.blinkdagger.androidLab.widget

import android.app.Activity
import android.graphics.drawable.ColorDrawable
import android.view.KeyEvent
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.PopupWindow
import androidx.annotation.AnimatorRes
import androidx.annotation.FloatRange
import androidx.annotation.LayoutRes
import androidx.annotation.StyleRes

/**
 * 类描述：通用PopupWindow
 * 创建人：ls
 * 创建时间：2017/12/7
 * 修改人：
 * 修改时间：
 * 修改备注：
 */
class CustomPopupWindow private constructor(contentView: View, width: Int, height: Int) : PopupWindow(contentView, width, height) {
    private var activity: Activity? = null
    private var onShowBgAlpha  = 0f //显示的时候背景的透明度
    private var onShowChangeBg  = false //显示的时候背景透明度是否变化
    private var onDismissChangeBg = false //消失的时候背景透明度是否变化

    fun setActivity(activity: Activity?) {
        this.activity = activity
    }

    fun setOnShowBgAlpha(@FloatRange(from = 0.0, to = 1.0) onShowBgAlpha: Float) {
        this.onShowBgAlpha = onShowBgAlpha
    }

    fun setOnShowChangeBg(onShowChangeBg: Boolean) {
        this.onShowChangeBg = onShowChangeBg
    }

    fun setOnDismissChangeBg(onDismissChangeBg: Boolean) {
        this.onDismissChangeBg = onDismissChangeBg
    }

    fun showChangeBg() {
        if (onShowChangeBg) {
            activity?.let {
                setBackgroundAlpha(it, onShowBgAlpha)
            }
        }
    }

    override fun showAtLocation(parent: View, gravity: Int, x: Int, y: Int) {
        showChangeBg()
        super.showAtLocation(parent, gravity, x, y)
    }

    override fun showAsDropDown(anchor: View, xoff: Int, yoff: Int, gravity: Int) {
        super.showAsDropDown(anchor, xoff, yoff, gravity)
        showChangeBg()
    }

    override fun dismiss() {
        if (onDismissChangeBg) {
            activity?.let {
                setBackgroundAlpha(it, 1.0f)
            }
        }
        super.dismiss()
    }

    class Builder(private val activity: Activity) {
        private var contentView // 内容视图
                : View? = null
        private var width = ViewGroup.LayoutParams.WRAP_CONTENT //宽度
        private var height = ViewGroup.LayoutParams.WRAP_CONTENT //高度
        @StyleRes
        private var animationStyle  = 0 // 动画
        private var focusable = false //是否可以获取焦点
        private var outsideTouchable  = false //外部是否可以点击
        private var onShowChangeBg = true //显示的时候背景透明度是否变透明【默认变化】
        private var onDismissChangeBg = true //消失的时候背景透明度是否还原【默认还原】
        private var onShowBgAlpha = 0.5f //显示的时候背景的透明度
        private var onKeyBackDissmiss = false //是否拦截返回键消失【默认不拦截】
        private var onDismissListener // 消失监听事件
                : OnDismissListener? = null

        fun setContentView(contentView: View?): Builder {
            this.contentView = contentView
            return this
        }

        fun setContentView(@LayoutRes contentViewId: Int): Builder {
            contentView = activity.layoutInflater.inflate(contentViewId, null)
            return this
        }

        fun setContentView(contentView: View?, width: Int, height: Int): Builder {
            this.width = width
            this.height = height
            return this
        }

        fun setContentView(@LayoutRes contentViewId: Int, width: Int, height: Int): Builder {
            contentView = activity.layoutInflater.inflate(contentViewId, null)
            this.width = width
            this.height = height
            return this
        }

        fun setAnimationStyle(@AnimatorRes animationStyle: Int): Builder {
            this.animationStyle = animationStyle
            return this
        }

        fun setFocusable(focusable: Boolean): Builder {
            this.focusable = focusable
            return this
        }

        fun setOutsideTouchable(outsideTouchable: Boolean): Builder {
            this.outsideTouchable = outsideTouchable
            return this
        }

        fun setOnShowChangeBg(onShowChangeBg: Boolean): Builder {
            this.onShowChangeBg = onShowChangeBg
            focusable = true
            return this
        }

        fun setOnDismissChangeBg(onDismissChangeBg: Boolean): Builder {
            this.onDismissChangeBg = onDismissChangeBg
            return this
        }

        fun setOnShowBgAlpha(@FloatRange(from = 0.0, to = 255.0) onShowBgAlpha: Float): Builder {
            this.onShowBgAlpha = onShowBgAlpha
            return this
        }

        fun setOnDismissListener(onDismissListener: OnDismissListener?) {
            this.onDismissListener = onDismissListener
        }

        /**
         * 设置点击外部本身是否消失
         *
         * @param onOutsideTouchDissmiss
         */
        fun setOnOutsideTouchDissmiss(onOutsideTouchDissmiss: Boolean): Builder {
            if (onOutsideTouchDissmiss) {
                outsideTouchable = true
                focusable = true
            } else {
                outsideTouchable = false
                focusable = false
            }
            return this
        }

        /**
         * 设置点击返回键本身是否消失
         *
         * @return
         */
        fun setOnKeyBackDissmiss(onKeyBackDissmiss: Boolean): Builder {
            this.onKeyBackDissmiss = onKeyBackDissmiss
            return this
        }

        fun build(): CustomPopupWindow {
            if (contentView == null) {
                throw NullPointerException("contentView con't be null ")
            }
            val popupWindow = CustomPopupWindow(contentView!!, width, height)
            popupWindow.setActivity(activity)
            popupWindow.isOutsideTouchable = outsideTouchable
            popupWindow.isFocusable = focusable
            popupWindow.setOnDismissChangeBg(onDismissChangeBg)
            popupWindow.setOnShowChangeBg(onShowChangeBg)
            popupWindow.setOnShowBgAlpha(onShowBgAlpha)
            popupWindow.setOnDismissListener(onDismissListener)
            popupWindow.setBackgroundDrawable(ColorDrawable())
            if (onKeyBackDissmiss) {
                contentView?.isFocusable = true // 这个很重要
                contentView?.isFocusableInTouchMode = true
                contentView?.setOnKeyListener(View.OnKeyListener { v, keyCode, event ->
                    if (keyCode == KeyEvent.KEYCODE_BACK) {
                        popupWindow.dismiss()
                        return@OnKeyListener true
                    }
                    false
                })
            }
            return popupWindow
        }

    }

    /**
     * 设置页面的透明度
     *
     * @param bgAlpha 1表示不透明
     */
    private fun setBackgroundAlpha(activity: Activity, bgAlpha: Float) {
        val lp = activity.window.attributes
        lp.alpha = bgAlpha
        if (bgAlpha == 1f) {
            activity.window.clearFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
        } else {
            activity.window.addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
        }
        activity.window.attributes = lp
    }
}